### ISSUE TITLE
plugin/dnssec: cache serves old RRSIGs after zone is changed

### ISSUE THREAD

**Issue by Isolus** [coredns/coredns#4448]
Since 4bf7347 after updating a zone (tested with file plugin and secondary plugin) coredns serves outdated cached RRSIGs (i.e. if you change the SOA the whole zone is rejected by validating resolvers). Since reverting the commit reintroduces #3953 I'm not sure how to handle this. Either the cache key would have to be extended again or the cache (or parts of it) would have to be reset when changing the zone.

**Comment by coelho** [comment]
This is a pretty critical issue as it causes downtime in a production environment that is DNSSEC enabled. 
Any updated zones fail to resolve due to incorrect RRSIGs.

**Comment by miekg** [comment]
current the cache doesn't check RRSIG expiration to decide if something should be cached. Anything within TTL *and* expiration should be valid.

I wonder though, why would you even have the cache plugin in the server block where you sign a zone?

**Comment by coelho** [comment]
@miekg 
This issue is reproducible with this config:

```
test.com {
        dnssec {
                key file /bind9/keys/test.com.private
        }
        auto {
                directory /bind9/zones db.(.*) {1}
		reload 2s
        }
	loadbalance round_robin
}
. {
	reload 2s
        auto {
                directory /bind9/zones db.(.*) {1}
		reload 2s
        }
	loadbalance round_robin
}
```

If you change the zone file, and it is reloaded, the generated RRSIG will be wrong.

This appears to be due to the CoreDNS's DNSSEC RRSIG caching, where the automatically signed zones get cached for performance reasons. It looks like it is not clearing the invalidating the cache properly anymore.

We reverted to v1.6.9 and we no longer have the issue*.

**Comment by miekg** [comment]
oh, OK _that_ cache. Yes that makes sense and is pretty dumb. Not sure why an older version works, because that cache is in ther since forever, but you claim the cache key is now done wrong.. Interesting

**Comment by miekg** [comment]
so either we use the entire record to get to a cache key which will work, but will litter the cache with older signatures that are not removed. And the `cache` plugin as-is lacks a method that get all keys so we iterate over it - maybe that should be added to remove older signatures...

**Comment by miekg** [comment]
maybe the whole caching was a mistake, it still vulnerable to ddossing I think, meaning it only adds value in the happy path, but then you still have this issue. No cache, no issue, but slower responses

**Comment by miekg** [comment]
can this thing live without a RRSIG cache? Or would that be a disaster?

**Comment by coelho** [comment]
@miekg from my understanding EC signing is very slow. QPS would be reduced (possibly dramatically), although I do not have solid numbers for how much.

**Comment by miekg** [comment]
how does https://github.com/coredns/coredns/pull/4537 work for you folks?

**Comment by Isolus** [comment]
For me, #4537 seems to have solved the issue and everything else seems to work.

**Comment by JensErat** [comment]
It seems to me that there is _another_ (not yet reported) issue fixed by the PR: Upgrading to 1.8.3 (from some old 1.6 release) broke issuing certificates with Let's Encrypt for me. They reported they received SRVFAIL when catching RAA records. I didn't not entirely check the DNS response, but requesting a non-existing RAA record against a validating resolver (ie. 8.8.8.8) also returned a SRVFAIL for me. Requesting directly against coredns was successful on the other hand.

Specific error message, in the hope Google catches up for others having the same issue with current CoreDNS release:

> Accepting challenge authorization failed: acme: authorization error for <domain> 400 urn:ietf:params:acme:error:dns: DNS problem: SERVFAIL looking up CAA for <domain> - the domain's nameservers may be malfunctioning

Running a build with #4537 included resulted in the expected NOERROR reply again, with expected authenticated data flag set.

I assume the cache fix also has some effect on NSEC records? This might make it an even worse issue, because it is even triggered without changes to the DNS zones, nor can it be mitigated by restarting CoreDNS.

The good side of the issue: I finally put CAA records on all domains...
